pyccel/codegen/printing/ccode.py

```diff
@@ -785,71 +776,89 @@ def _print_MathFunctionBase(self, expr):
 
         """
         # add necessary include
-        self._additional_imports.add('math')
         type_name = type(expr).__name__
         try:
             func_name = math_function_to_c[type_name]
         except KeyError:
             errors.report(PYCCEL_RESTRICTION_TODO, severity='fatal')
+
+        if func_name.startswith("pyc"):
+            self._additional_imports.add('pyc_math')
+        else:
+            if expr.dtype is NativeComplex():
+                self._additional_imports.add('cmath')
+            else:
+                self._additional_imports.add('math')
         args = []
         for arg in expr.args:
-            if arg.dtype is NativeComplex():
-                self._additional_imports.add('complex')
-            if arg.dtype is not NativeReal():
-                args.append(self._print(NumpyFloat(arg)))
-            else :
+            if arg.dtype != expr.dtype:
+                cast_func = python_builtin_datatypes[str_dtype(expr.dtype)]
+                args.append(self._print(cast_func(arg)))
+            else:
                 args.append(self._print(arg))
```

Comment:
Just checking. Does this mean that all math functions (except 1 or 2 exceptions such as MathFloor) have the same output type as their arguments? That seems like a good way of looking at it :)

Reply:
yes, Most of them are returning the same type of inputs.
for the others, we create a specific printer for them.